File: backend/src/app/schemas/social.py
"""Schemas for social endpoints (Phase 9)."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from src.app.models.enums import WatchStatus


class SocialFeedItemResponse(BaseModel):
    """Single activity item in group social feed."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    entry_id: UUID
    user_id: UUID
    media_id: UUID
    event_type: str
    old_status: WatchStatus | None = None
    new_status: WatchStatus | None = None
    old_progress: int | None = None
    new_progress: int | None = None
    old_score: float | None = None
    new_score: float | None = None
    note: str | None = None
    created_at: datetime


class SocialFeedResponse(BaseModel):
    """Paginated group activity feed response."""

    items: list[SocialFeedItemResponse]
    total: int
    limit: int
    offset: int


class RecommendationCreateRequest(BaseModel):
    """Payload to recommend a media title to another user."""

    to_user_id: UUID
    media_id: UUID
    message: str | None = None


class RecommendationResponse(BaseModel):
    """Recommendation response model."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    from_user_id: UUID
    to_user_id: UUID
    media_id: UUID
    message: str | None = None
    is_acknowledged: bool
    acknowledged_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class RecommendationInboxResponse(BaseModel):
    """Paginated inbox recommendations for current user."""

    items: list[RecommendationResponse]
    total: int
    limit: int
    offset: int


class RecommendationSentResponse(BaseModel):
    """Paginated sent recommendations for current user."""

    items: list[RecommendationResponse]
    total: int
    limit: int
    offset: int


class DiscussionCreateRequest(BaseModel):
    """Payload to create a discussion thread."""

    media_id: UUID
    group_id: UUID
    title: str | None = None
    body: str
    has_spoilers: bool = False
    episode_number: int | None = None
    chapter_number: float | None = None


class DiscussionResponse(BaseModel):
    """Discussion response model."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    media_id: UUID
    group_id: UUID
    user_id: UUID
    title: str | None = None
    body: str
    has_spoilers: bool
    episode_number: int | None = None
    chapter_number: float | None = None
    created_at: datetime
    updated_at: datetime


class DiscussionListResponse(BaseModel):
    """Paginated discussion list for a media entry."""

    items: list[DiscussionResponse]
    total: int
    limit: int
    offset: int


class DiscussionReplyCreateRequest(BaseModel):
    """Payload to create a reply under a discussion."""

    body: str
    has_spoilers: bool = False
    parent_reply_id: UUID | None = None


class DiscussionReplyResponse(BaseModel):
    """Discussion reply response model."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    discussion_id: UUID
    user_id: UUID
    parent_reply_id: UUID | None = None
    body: str
    has_spoilers: bool
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/social.ts
/** Social feed activity item from list_entry_history */
export interface FeedActivityItem {
  id: string
  entry_id: string
  user_id: string
  media_id: string
  event_type: string
  old_status?: string | null
  new_status?: string | null
  old_progress?: number | null
  new_progress?: number | null
  old_score?: number | null
  new_score?: number | null
  note?: string | null
  created_at: string
}

export interface SocialFeedResponse {
  items: FeedActivityItem[]
  total: number
  limit: number
  offset: number
}

/** Recommendation from another group member */
export interface Recommendation {
  id: string
  from_user_id: string
  to_user_id: string
  media_id: string
  message?: string | null
  is_acknowledged: boolean
  acknowledged_at?: string | null
  created_at: string
  updated_at: string
}

export interface RecommendationInboxResponse {
  items: Recommendation[]
  total: number
  limit: number
  offset: number
}

export interface RecommendationSentResponse {
  items: Recommendation[]
  total: number
  limit: number
  offset: number
}

export interface RecommendationCreateRequest {
  to_user_id: string
  media_id: string
  message?: string | null
}

/** Discussion thread for a media entry */
export interface Discussion {
  id: string
  media_id: string
  group_id: string
  user_id: string
  title?: string | null
  body: string
  has_spoilers: boolean
  episode_number?: number | null
  chapter_number?: number | null
  created_at: string
  updated_at: string
}

export interface DiscussionListResponse {
  items: Discussion[]
  total: number
  limit: number
  offset: number
}

export interface DiscussionCreateRequest {
  media_id: string
  group_id: string
  title?: string | null
  body: string
  has_spoilers?: boolean
  episode_number?: number | null
  chapter_number?: number | null
}

/** Reply under a discussion thread */
export interface DiscussionReply {
  id: string
  discussion_id: string
  user_id: string
  parent_reply_id?: string | null
  body: string
  has_spoilers: boolean
  created_at: string
  updated_at: string
}

export interface DiscussionReplyListResponse {
  items: DiscussionReply[]
  total: number
  limit: number
  offset: number
}

export interface DiscussionReplyCreateRequest {
  body: string
  has_spoilers?: boolean
  parent_reply_id?: string | null
}

/** Generic paginated response helper */
export interface PaginatedResponse<T> {
  items: T[]
  total: number
  limit: number
  offset: number
}

/** User profile (for profile page) */
export interface UserProfile {
  id: string
  username: string
  display_name?: string | null
  email: string
  avatar_url?: string | null
  bio?: string | null
  timezone: string
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface UserProfileUpdateRequest {
  display_name?: string | null
  avatar_url?: string | null
  bio?: string | null
  timezone?: string | null
}

export interface PasswordChangeRequest {
  current_password: string
  new_password: string
}

/** User settings (theme/language/timezone) */
export interface UserSettings {
  user_id: string
  theme: string
  language: string
  timezone: string
  created_at: string
  updated_at: string
}

export interface UserSettingsUpdateRequest {
  theme?: string
  language?: string
  timezone?: string
}
